### ISSUE TITLE
network - Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory  

### ISSUE THREAD

**Issue by pakrym** [Azure/autorest.modelerfour#285]
```
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xa0e670 node::Abort() [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 2: 0xa0ea9c node::OnFatalError(char const*, char const*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 3: 0xb83afe v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 4: 0xb83e79 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 5: 0xd32305  [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 6: 0xd32996 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 7: 0xd41209 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 8: 0xd42045 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 9: 0xd44b1c v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
10: 0xd0bd22 v8::internal::Handle<v8::internal::FixedArray> v8::internal::Factory::NewFixedArrayWithMap<v8::internal::FixedArray>(v8::internal::RootIndex, int, v8::internal::AllocationType) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
11: 0xf2c4b9 v8::internal::HashTable<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::New(v8::internal::Isolate*, int, v8::internal::AllocationType, v8::internal::MinimumCapacity) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
12: 0xf34c3e v8::internal::Dictionary<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::Add(v8::internal::Isolate*, v8::internal::Handle<v8::internal::NameDictionary>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDetails, int*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
13: 0xf34db9 v8::internal::BaseNameDictionary<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::AddNoUpdateNextEnumerationIndex(v8::internal::Isolate*, v8::internal::Handle<v8::internal::NameDictionary>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDetails, int*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
14: 0xf37313 v8::internal::BaseNameDictionary<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::Add(v8::internal::Isolate*, v8::internal::Handle<v8::internal::NameDictionary>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDetails, int*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
15: 0xefff72 v8::internal::LookupIterator::ApplyTransitionToDataProperty(v8::internal::Handle<v8::internal::JSReceiver>) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
16: 0xf314b1 v8::internal::Object::AddDataProperty(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::Maybe<v8::internal::ShouldThrow>, v8::internal::StoreOrigin) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
17: 0xedb110 v8::internal::JSObject::DefineOwnPropertyIgnoreAttributes(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::Maybe<v8::internal::ShouldThrow>, v8::internal::JSObject::AccessorInfoHandling) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
18: 0xedb349 v8::internal::JSObject::DefineOwnPropertyIgnoreAttributes(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::internal::JSObject::AccessorInfoHandling) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
19: 0xe21ed0 v8::internal::JsonParser<unsigned short>::BuildJsonObject(v8::internal::JsonParser<unsigned short>::JsonContinuation const&, std::vector<v8::internal::JsonProperty, std::allocator<v8::internal::JsonProperty> > const&, v8::internal::Handle<v8::internal::Map>) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
20: 0xe22a83 v8::internal::JsonParser<unsigned short>::ParseJsonValue() [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
21: 0xe2393f v8::internal::JsonParser<unsigned short>::ParseJson() [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
22: 0xc36f05 v8::internal::Builtin_JsonParse(int, unsigned long*, v8::internal::Isolate*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
23: 0x13da5f9  [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
Process() cancelled due to failure 
Invoke: /home/vsts/work/1/s/eng/Generation.psm1:45
Line |
  45 |      Invoke $command
     |      ~~~~~~~~~~~~~~~
     | Command failed to execute: /home/vsts/work/1/s/node_modules/.bin/autorest-beta --require=/home/vsts/work/1/s/readme.md https://github.com/Azure/azure-rest-api-specs/blob/d9cf7c7ed3d674ebd482836e82b274014245ae67/specification/network/resource-manager/readme.md --namespace=network_resource_manager --output-folder=/home/vsts/work/1/s/samples/smoketests/network-resource-manager/network-resource-manager

```

**Comment by pakrym** [comment]
Same issue with cosmos.

**Comment by jhendrixMSFT** [comment]
I suspect that some of this is related to `PreCheck/AllOfWhenYouMeantRef`.  I guess this was recently enabled?  At any rate, before this change I could build one network tag in ~3 minutes.  Now it takes a _really long time_ and eventually crashes.

**Comment by daviwil** [comment]
PR #284 fixes this issue, it's now been released in `@autorest/modelerfour` [4.13.351](https://github.com/Azure/autorest.modelerfour/releases/tag/v4.13.351).

**Comment by pakrym** [comment]
I still see this failure for web on CI but not locally.

**Comment by daviwil** [comment]
The same YAML failure? And you're using the newest m4 on CI too?

**Comment by pakrym** [comment]
Same out of memory issue,

```
Time Elapsed 00:00:03.51
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xa0e670 node::Abort() [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 2: 0xa0ea9c node::OnFatalError(char const*, char const*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 3: 0xb83afe v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 4: 0xb83e79 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 5: 0xd32305  [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 6: 0xd32996 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 7: 0xd41209 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 8: 0xd42045 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 9: 0xd44b1c v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
10: 0xd0a2cd  [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
11: 0xd0b352 v8::internal::Factory::AllocateRawWithAllocationSite(v8::internal::Handle<v8::internal::Map>, v8::internal::AllocationType, v8::internal::Handle<v8::internal::AllocationSite>) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
12: 0xd14ae4 v8::internal::Factory::NewJSArrayWithUnverifiedElements(v8::internal::Handle<v8::internal::FixedArrayBase>, v8::internal::ElementsKind, int, v8::internal::AllocationType) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
13: 0xd14d50 v8::internal::Factory::NewJSArray(v8::internal::ElementsKind, int, int, v8::internal::ArrayStorageAllocationMode, v8::internal::AllocationType) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
14: 0xe1a951 v8::internal::JsonParser<unsigned short>::BuildJsonArray(v8::internal::JsonParser<unsigned short>::JsonContinuation const&, std::vector<v8::internal::Handle<v8::internal::Object>, std::allocator<v8::internal::Handle<v8::internal::Object> > > const&) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
15: 0xe231e2 v8::internal::JsonParser<unsigned short>::ParseJsonValue() [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
16: 0xe2393f v8::internal::JsonParser<unsigned short>::ParseJson() [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
17: 0xc36f05 v8::internal::Builtin_JsonParse(int, unsigned long*, v8::internal::Isolate*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
18: 0x13da5f9  [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
``
Yeah, it's on latest, I'm digging to get more details.

**Comment by daviwil** [comment]
Which version of Node on CI? Also, where was AutoRest in the pipeline at that time? Must have been early of its talking about JSON parsing.

**Comment by pakrym** [comment]
`AutoRest code generation utility [cli version: 3.0.6222; node: v13.14.0, max-memory: 8192 gb]`

Issue might be caused by us running in parallel in language constrained environment.

**Comment by pakrym** [comment]
Nope, single threaded didn't help, switching to windows didn't help but moving it to it's own run did help.

I would keep this issue open so we can look at the memory usage issues later.

**Comment by pakrym** [comment]
I take it back, it's wasn't `web` but rather `SQL` and it crashes locally too.

**Comment by daviwil** [comment]
I can confirm that I see the crash with SQL too, I think that some JSON serialization we're doing in Precheck is triggering it.

**Comment by daviwil** [comment]
PR #286 addresses the heap crash issue, @pakrym has confirmed that the test build is working for SQL.  Will post an update here once it's been merged and a new release with the fix is out.

**Comment by daviwil** [comment]
Fix for this issue is included in the just-released `4.13.358`.
